agents/dqn.py

```diff
@@ -1,137 +1,132 @@
-
-
 class DQNAgent(TorchAgent, config_name='dqn'):
     def __init__(
             self,
             current_node_idx_prefix: str,
             destination_node_idx_prefix: str,
             neighbors_node_ids_prefix: str,
+            output_prefix: str,
+            bag_ids_prefix: str,
             q_network: BaseQNetwork,
             discount_factor: float = 0.99,
             research_prob: float = 0.1,
             bag_trajectory_memory: BaseBagTrajectoryMemory = None,
-            node_idx: int = None,
-            sample_size: int = 100,
+            trajectory_sample_size: int = 30,
+            optimizer_factory: Callable[[nn.Module], BaseOptimizer] = None
     ):
         assert 0 < discount_factor < 1, 'Incorrect `discount_factor` choice'
         assert 0 < research_prob < 1, 'Incorrect `discount_factor` choice'
         super().__init__()
 
+        self._node_id = -1
+
         self._current_node_idx_prefix = current_node_idx_prefix
         self._destination_node_idx_prefix = destination_node_idx_prefix
         self._neighbors_node_ids_prefix = neighbors_node_ids_prefix
+        self._output_prefix = output_prefix
+        self._bag_ids_prefix = bag_ids_prefix
 
-        self._node_idx = node_idx
         self._q_network = q_network
         self._discount_factor = discount_factor
         self._research_prob = research_prob
         self._bag_trajectory_memory = bag_trajectory_memory
-        self._sample_size = sample_size
+        self._trajectory_sample_size = trajectory_sample_size
+
+        self._optimizer = optimizer_factory(self) if optimizer_factory is not None else None
 
     @classmethod
     def create_from_config(cls, config):
         return cls(
             current_node_idx_prefix=config['current_node_idx_prefix'],
             destination_node_idx_prefix=config['destination_node_idx_prefix'],
             neighbors_node_ids_prefix=config['neighbors_node_ids_prefix'],
+            output_prefix=config['output_prefix'],
+            bag_ids_prefix=config.get('bag_ids_prefix', None),
             q_network=BaseEncoder.create_from_config(config['q_network']),
             discount_factor=config.get('discount_factor', 0.99),
             research_prob=config.get('research_prob', 0.1),
             bag_trajectory_memory=BaseBagTrajectoryMemory.create_from_config(config['path_memory'])
             if 'path_memory' in config else None,
-            node_idx=config.get('node_idx', None),
-            sample_size=config.get('sample_size', 100)
+            trajectory_sample_size=config.get('trajectory_sample_size', 30),
+            optimizer_factory=lambda m: BaseOptimizer.create_from_config(config['optimizer'], model=m)
+            if 'optimizer' in config else None
         )
 
     def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
         # Shape: [batch_size]
         current_node_idx = inputs[self._current_node_idx_prefix]
+        if self._node_id < 0:
```

Comment:
I think it can be done in the `create_from_config` function and `__init__` function by adding additional parameter in the constructor and function

Reply:
Sasha makes deep copy over this agents in simulation, is it good idea?)

Reply:
In such case, it should be re-assigned somewhere. Just make this field non-private (remove `_` prefix) since now it is referenced outside the object.

Reply:
Lets stay in this way yet